<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_14) on Fri Sep 18 14:09:09 BST 2009 -->
<TITLE>
NCBIBlastInputParameters
</TITLE>

<META NAME="date" CONTENT="2009-09-18">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="NCBIBlastInputParameters";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/NCBIBlastInputParameters.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInput.html" title="class in uk.org.mygrid.cagrid.domain.ncbiblast"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastJob.html" title="class in uk.org.mygrid.cagrid.domain.ncbiblast"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="NCBIBlastInputParameters.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
uk.org.mygrid.cagrid.domain.ncbiblast</FONT>
<BR>
Class NCBIBlastInputParameters</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../../../resources/inherit.gif" ALT="extended by "><B>uk.org.mygrid.cagrid.domain.ncbiblast.NCBIBlastInputParameters</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.io.Serializable</DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>NCBIBlastInputParameters</B><DT>extends java.lang.Object<DT>implements java.io.Serializable</DL>
</PRE>

<P>
Input parameters for the NCBI BLAST job.
<P>

<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../../../../serialized-form.html#uk.org.mygrid.cagrid.domain.ncbiblast.NCBIBlastInputParameters">Serialized Form</A></DL>
<HR>

<P>

<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#NCBIBlastInputParameters()">NCBIBlastInputParameters</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#NCBIBlastInputParameters(java.math.BigInteger, uk.org.mygrid.cagrid.valuedomains.BLASTProgram, java.math.BigInteger, java.lang.String, java.lang.Double, java.math.BigInteger, java.lang.Boolean, java.lang.Boolean, java.math.BigInteger, uk.org.mygrid.cagrid.valuedomains.Matrix, java.math.BigInteger, java.math.BigInteger, java.math.BigInteger, uk.org.mygrid.cagrid.domain.common.MolecularSequenceDatabase)">NCBIBlastInputParameters</A></B>(java.math.BigInteger&nbsp;alignmentsToOutput,
                         <A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/BLASTProgram.html" title="class in uk.org.mygrid.cagrid.valuedomains">BLASTProgram</A>&nbsp;blastProgram,
                         java.math.BigInteger&nbsp;dropoff,
                         java.lang.String&nbsp;email,
                         java.lang.Double&nbsp;expectedThreshold,
                         java.math.BigInteger&nbsp;extendGap,
                         java.lang.Boolean&nbsp;filter,
                         java.lang.Boolean&nbsp;gapAlignment,
                         java.math.BigInteger&nbsp;match,
                         <A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/Matrix.html" title="class in uk.org.mygrid.cagrid.valuedomains">Matrix</A>&nbsp;matrix,
                         java.math.BigInteger&nbsp;maxScores,
                         java.math.BigInteger&nbsp;mismatch,
                         java.math.BigInteger&nbsp;openGap,
                         <A HREF="../../../../../../uk/org/mygrid/cagrid/domain/common/MolecularSequenceDatabase.html" title="class in uk.org.mygrid.cagrid.domain.common">MolecularSequenceDatabase</A>&nbsp;queryDatabase)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#equals(java.lang.Object)">equals</A></B>(java.lang.Object&nbsp;obj)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.math.BigInteger</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getAlignmentsToOutput()">getAlignmentsToOutput</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the alignmentsToOutput value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/BLASTProgram.html" title="class in uk.org.mygrid.cagrid.valuedomains">BLASTProgram</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getBlastProgram()">getBlastProgram</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the blastProgram value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;org.apache.axis.encoding.Deserializer</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getDeserializer(java.lang.String, java.lang.Class, javax.xml.namespace.QName)">getDeserializer</A></B>(java.lang.String&nbsp;mechType,
                java.lang.Class&nbsp;_javaType,
                javax.xml.namespace.QName&nbsp;_xmlType)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get Custom Deserializer</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.math.BigInteger</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getDropoff()">getDropoff</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the dropoff value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getEmail()">getEmail</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the email value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getExpectedThreshold()">getExpectedThreshold</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the expectedThreshold value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.math.BigInteger</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getExtendGap()">getExtendGap</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the extendGap value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getFilter()">getFilter</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the filter value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getGapAlignment()">getGapAlignment</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the gapAlignment value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.math.BigInteger</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getMatch()">getMatch</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the match value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/Matrix.html" title="class in uk.org.mygrid.cagrid.valuedomains">Matrix</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getMatrix()">getMatrix</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the matrix value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.math.BigInteger</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getMaxScores()">getMaxScores</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the maxScores value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.math.BigInteger</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getMismatch()">getMismatch</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the mismatch value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.math.BigInteger</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getOpenGap()">getOpenGap</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the openGap value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../../../uk/org/mygrid/cagrid/domain/common/MolecularSequenceDatabase.html" title="class in uk.org.mygrid.cagrid.domain.common">MolecularSequenceDatabase</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getQueryDatabase()">getQueryDatabase</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets the queryDatabase value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;org.apache.axis.encoding.Serializer</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getSerializer(java.lang.String, java.lang.Class, javax.xml.namespace.QName)">getSerializer</A></B>(java.lang.String&nbsp;mechType,
              java.lang.Class&nbsp;_javaType,
              javax.xml.namespace.QName&nbsp;_xmlType)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get Custom Serializer</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;org.apache.axis.description.TypeDesc</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#getTypeDesc()">getTypeDesc</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return type metadata object</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#hashCode()">hashCode</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setAlignmentsToOutput(java.math.BigInteger)">setAlignmentsToOutput</A></B>(java.math.BigInteger&nbsp;alignmentsToOutput)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the alignmentsToOutput value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setBlastProgram(uk.org.mygrid.cagrid.valuedomains.BLASTProgram)">setBlastProgram</A></B>(<A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/BLASTProgram.html" title="class in uk.org.mygrid.cagrid.valuedomains">BLASTProgram</A>&nbsp;blastProgram)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the blastProgram value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setDropoff(java.math.BigInteger)">setDropoff</A></B>(java.math.BigInteger&nbsp;dropoff)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the dropoff value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setEmail(java.lang.String)">setEmail</A></B>(java.lang.String&nbsp;email)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the email value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setExpectedThreshold(java.lang.Double)">setExpectedThreshold</A></B>(java.lang.Double&nbsp;expectedThreshold)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the expectedThreshold value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setExtendGap(java.math.BigInteger)">setExtendGap</A></B>(java.math.BigInteger&nbsp;extendGap)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the extendGap value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setFilter(java.lang.Boolean)">setFilter</A></B>(java.lang.Boolean&nbsp;filter)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the filter value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setGapAlignment(java.lang.Boolean)">setGapAlignment</A></B>(java.lang.Boolean&nbsp;gapAlignment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the gapAlignment value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setMatch(java.math.BigInteger)">setMatch</A></B>(java.math.BigInteger&nbsp;match)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the match value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setMatrix(uk.org.mygrid.cagrid.valuedomains.Matrix)">setMatrix</A></B>(<A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/Matrix.html" title="class in uk.org.mygrid.cagrid.valuedomains">Matrix</A>&nbsp;matrix)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the matrix value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setMaxScores(java.math.BigInteger)">setMaxScores</A></B>(java.math.BigInteger&nbsp;maxScores)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the maxScores value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setMismatch(java.math.BigInteger)">setMismatch</A></B>(java.math.BigInteger&nbsp;mismatch)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the mismatch value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setOpenGap(java.math.BigInteger)">setOpenGap</A></B>(java.math.BigInteger&nbsp;openGap)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the openGap value for this NCBIBlastInputParameters.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html#setQueryDatabase(uk.org.mygrid.cagrid.domain.common.MolecularSequenceDatabase)">setQueryDatabase</A></B>(<A HREF="../../../../../../uk/org/mygrid/cagrid/domain/common/MolecularSequenceDatabase.html" title="class in uk.org.mygrid.cagrid.domain.common">MolecularSequenceDatabase</A>&nbsp;queryDatabase)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the queryDatabase value for this NCBIBlastInputParameters.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>getClass, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="NCBIBlastInputParameters()"><!-- --></A><H3>
NCBIBlastInputParameters</H3>
<PRE>
public <B>NCBIBlastInputParameters</B>()</PRE>
<DL>
</DL>
<HR>

<A NAME="NCBIBlastInputParameters(java.math.BigInteger, uk.org.mygrid.cagrid.valuedomains.BLASTProgram, java.math.BigInteger, java.lang.String, java.lang.Double, java.math.BigInteger, java.lang.Boolean, java.lang.Boolean, java.math.BigInteger, uk.org.mygrid.cagrid.valuedomains.Matrix, java.math.BigInteger, java.math.BigInteger, java.math.BigInteger, uk.org.mygrid.cagrid.domain.common.MolecularSequenceDatabase)"><!-- --></A><H3>
NCBIBlastInputParameters</H3>
<PRE>
public <B>NCBIBlastInputParameters</B>(java.math.BigInteger&nbsp;alignmentsToOutput,
                                <A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/BLASTProgram.html" title="class in uk.org.mygrid.cagrid.valuedomains">BLASTProgram</A>&nbsp;blastProgram,
                                java.math.BigInteger&nbsp;dropoff,
                                java.lang.String&nbsp;email,
                                java.lang.Double&nbsp;expectedThreshold,
                                java.math.BigInteger&nbsp;extendGap,
                                java.lang.Boolean&nbsp;filter,
                                java.lang.Boolean&nbsp;gapAlignment,
                                java.math.BigInteger&nbsp;match,
                                <A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/Matrix.html" title="class in uk.org.mygrid.cagrid.valuedomains">Matrix</A>&nbsp;matrix,
                                java.math.BigInteger&nbsp;maxScores,
                                java.math.BigInteger&nbsp;mismatch,
                                java.math.BigInteger&nbsp;openGap,
                                <A HREF="../../../../../../uk/org/mygrid/cagrid/domain/common/MolecularSequenceDatabase.html" title="class in uk.org.mygrid.cagrid.domain.common">MolecularSequenceDatabase</A>&nbsp;queryDatabase)</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="getBlastProgram()"><!-- --></A><H3>
getBlastProgram</H3>
<PRE>
public <A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/BLASTProgram.html" title="class in uk.org.mygrid.cagrid.valuedomains">BLASTProgram</A> <B>getBlastProgram</B>()</PRE>
<DL>
<DD>Gets the blastProgram value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>blastProgram BLAST program to use. Valid values are: BLASTN, BLASTP and BLASTX.</DL>
</DD>
</DL>
<HR>

<A NAME="setBlastProgram(uk.org.mygrid.cagrid.valuedomains.BLASTProgram)"><!-- --></A><H3>
setBlastProgram</H3>
<PRE>
public void <B>setBlastProgram</B>(<A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/BLASTProgram.html" title="class in uk.org.mygrid.cagrid.valuedomains">BLASTProgram</A>&nbsp;blastProgram)</PRE>
<DL>
<DD>Sets the blastProgram value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>blastProgram</CODE> - BLAST program to use. Valid values are: BLASTN, BLASTP and BLASTX.</DL>
</DD>
</DL>
<HR>

<A NAME="getAlignmentsToOutput()"><!-- --></A><H3>
getAlignmentsToOutput</H3>
<PRE>
public java.math.BigInteger <B>getAlignmentsToOutput</B>()</PRE>
<DL>
<DD>Gets the alignmentsToOutput value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>alignmentsToOutput Number of database sequences to output alignments for.</DL>
</DD>
</DL>
<HR>

<A NAME="setAlignmentsToOutput(java.math.BigInteger)"><!-- --></A><H3>
setAlignmentsToOutput</H3>
<PRE>
public void <B>setAlignmentsToOutput</B>(java.math.BigInteger&nbsp;alignmentsToOutput)</PRE>
<DL>
<DD>Sets the alignmentsToOutput value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>alignmentsToOutput</CODE> - Number of database sequences to output alignments for.</DL>
</DD>
</DL>
<HR>

<A NAME="getDropoff()"><!-- --></A><H3>
getDropoff</H3>
<PRE>
public java.math.BigInteger <B>getDropoff</B>()</PRE>
<DL>
<DD>Gets the dropoff value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>dropoff Amount a score must drop before extension of word hits is halted.</DL>
</DD>
</DL>
<HR>

<A NAME="setDropoff(java.math.BigInteger)"><!-- --></A><H3>
setDropoff</H3>
<PRE>
public void <B>setDropoff</B>(java.math.BigInteger&nbsp;dropoff)</PRE>
<DL>
<DD>Sets the dropoff value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>dropoff</CODE> - Amount a score must drop before extension of word hits is halted.</DL>
</DD>
</DL>
<HR>

<A NAME="getEmail()"><!-- --></A><H3>
getEmail</H3>
<PRE>
public java.lang.String <B>getEmail</B>()</PRE>
<DL>
<DD>Gets the email value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>email E-mail address of the user.</DL>
</DD>
</DL>
<HR>

<A NAME="setEmail(java.lang.String)"><!-- --></A><H3>
setEmail</H3>
<PRE>
public void <B>setEmail</B>(java.lang.String&nbsp;email)</PRE>
<DL>
<DD>Sets the email value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>email</CODE> - E-mail address of the user.</DL>
</DD>
</DL>
<HR>

<A NAME="getExpectedThreshold()"><!-- --></A><H3>
getExpectedThreshold</H3>
<PRE>
public java.lang.Double <B>getExpectedThreshold</B>()</PRE>
<DL>
<DD>Gets the expectedThreshold value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>expectedThreshold The expected threshold establishes a statistical significance threshold
 for reporting database sequence matches. The default value is 10,
 meaning that 10 matches are expected to be found merely by chance.
 Lower expected thresholds are more stringent, leading to fewer chance
 matches being reported. Increasing the expected threshold shows less
 stringent matches and is recommended when one is performing searches
 with short sequences as a short query is more likely to occur by chance
 in the database than a longer one, so even a perfect match (no gaps)
 can have low statistical significance and may not be reported. Increasing
 the expected threshold allows you to look further down in the hit
 list and see matches that would normally be discarded because of low
 statistical significance. Generally a value of up to 1000 is enough
 to see results.</DL>
</DD>
</DL>
<HR>

<A NAME="setExpectedThreshold(java.lang.Double)"><!-- --></A><H3>
setExpectedThreshold</H3>
<PRE>
public void <B>setExpectedThreshold</B>(java.lang.Double&nbsp;expectedThreshold)</PRE>
<DL>
<DD>Sets the expectedThreshold value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>expectedThreshold</CODE> - The expected threshold establishes a statistical significance threshold
 for reporting database sequence matches. The default value is 10,
 meaning that 10 matches are expected to be found merely by chance.
 Lower expected thresholds are more stringent, leading to fewer chance
 matches being reported. Increasing the expected threshold shows less
 stringent matches and is recommended when one is performing searches
 with short sequences as a short query is more likely to occur by chance
 in the database than a longer one, so even a perfect match (no gaps)
 can have low statistical significance and may not be reported. Increasing
 the expected threshold allows you to look further down in the hit
 list and see matches that would normally be discarded because of low
 statistical significance. Generally a value of up to 1000 is enough
 to see results.</DL>
</DD>
</DL>
<HR>

<A NAME="getExtendGap()"><!-- --></A><H3>
getExtendGap</H3>
<PRE>
public java.math.BigInteger <B>getExtendGap</B>()</PRE>
<DL>
<DD>Gets the extendGap value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>extendGap Gap extension penalty that is added to the standard gap open penalty
 for each base or residue in the gap. This is how long gaps are penalised
 - if you don't want long gaps, increase the extension gap penalty.
 Usually one can expect a few long gaps rather than many short gaps,
 so the gap extension penalty should be lower than the gap penalty.
 An exception is where one or both sequences are single reads with
 possible sequencing errors in which case one would expect many single
 base gaps. One can get this result by setting the gap open penalty
 to zero (or very low) and using the gap extension penalty to control
 gap scoring.</DL>
</DD>
</DL>
<HR>

<A NAME="setExtendGap(java.math.BigInteger)"><!-- --></A><H3>
setExtendGap</H3>
<PRE>
public void <B>setExtendGap</B>(java.math.BigInteger&nbsp;extendGap)</PRE>
<DL>
<DD>Sets the extendGap value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>extendGap</CODE> - Gap extension penalty that is added to the standard gap open penalty
 for each base or residue in the gap. This is how long gaps are penalised
 - if you don't want long gaps, increase the extension gap penalty.
 Usually one can expect a few long gaps rather than many short gaps,
 so the gap extension penalty should be lower than the gap penalty.
 An exception is where one or both sequences are single reads with
 possible sequencing errors in which case one would expect many single
 base gaps. One can get this result by setting the gap open penalty
 to zero (or very low) and using the gap extension penalty to control
 gap scoring.</DL>
</DD>
</DL>
<HR>

<A NAME="getFilter()"><!-- --></A><H3>
getFilter</H3>
<PRE>
public java.lang.Boolean <B>getFilter</B>()</PRE>
<DL>
<DD>Gets the filter value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>filter The filter allows one to mask out various segments of the query
 sequence for regions which are non-specific for sequence similarity
 searches. Filtering can eliminate statistically significant but biologically
 uninteresting reports from the output, for example hits against common
 acidic-, basic- or proline-rich regions, leaving the more biologically
 interesting regions of the query sequence available for specific matching
 against database sequences. Filtering is only applied to the query
 sequence, not to database sequences. The program used for this, with
 nucleotide query sequences is known as DUST. The SEG program is used
 for filtering low complexity regions in amino acid sequences from
 one's protein query sequence. The default is false for protein searches
 and true for nucleotide searches.</DL>
</DD>
</DL>
<HR>

<A NAME="setFilter(java.lang.Boolean)"><!-- --></A><H3>
setFilter</H3>
<PRE>
public void <B>setFilter</B>(java.lang.Boolean&nbsp;filter)</PRE>
<DL>
<DD>Sets the filter value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>filter</CODE> - The filter allows one to mask out various segments of the query
 sequence for regions which are non-specific for sequence similarity
 searches. Filtering can eliminate statistically significant but biologically
 uninteresting reports from the output, for example hits against common
 acidic-, basic- or proline-rich regions, leaving the more biologically
 interesting regions of the query sequence available for specific matching
 against database sequences. Filtering is only applied to the query
 sequence, not to database sequences. The program used for this, with
 nucleotide query sequences is known as DUST. The SEG program is used
 for filtering low complexity regions in amino acid sequences from
 one's protein query sequence. The default is false for protein searches
 and true for nucleotide searches.</DL>
</DD>
</DL>
<HR>

<A NAME="getGapAlignment()"><!-- --></A><H3>
getGapAlignment</H3>
<PRE>
public java.lang.Boolean <B>getGapAlignment</B>()</PRE>
<DL>
<DD>Gets the gapAlignment value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>gapAlignment Whether to perform optimised alignments within regions involving
 gaps. If set to true, the program will perform an alignment using
 gaps. Otherwise, it will report only individual HSP (High-Scoring
 Segment Pair) where two sequence match each other, and thus will not
 produce alignments with gaps.</DL>
</DD>
</DL>
<HR>

<A NAME="setGapAlignment(java.lang.Boolean)"><!-- --></A><H3>
setGapAlignment</H3>
<PRE>
public void <B>setGapAlignment</B>(java.lang.Boolean&nbsp;gapAlignment)</PRE>
<DL>
<DD>Sets the gapAlignment value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>gapAlignment</CODE> - Whether to perform optimised alignments within regions involving
 gaps. If set to true, the program will perform an alignment using
 gaps. Otherwise, it will report only individual HSP (High-Scoring
 Segment Pair) where two sequence match each other, and thus will not
 produce alignments with gaps.</DL>
</DD>
</DL>
<HR>

<A NAME="getMatch()"><!-- --></A><H3>
getMatch</H3>
<PRE>
public java.math.BigInteger <B>getMatch</B>()</PRE>
<DL>
<DD>Gets the match value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>match Reward for a nucleotide match. Only when program is set to BLASTN.</DL>
</DD>
</DL>
<HR>

<A NAME="setMatch(java.math.BigInteger)"><!-- --></A><H3>
setMatch</H3>
<PRE>
public void <B>setMatch</B>(java.math.BigInteger&nbsp;match)</PRE>
<DL>
<DD>Sets the match value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>match</CODE> - Reward for a nucleotide match. Only when program is set to BLASTN.</DL>
</DD>
</DL>
<HR>

<A NAME="getMatrix()"><!-- --></A><H3>
getMatrix</H3>
<PRE>
public <A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/Matrix.html" title="class in uk.org.mygrid.cagrid.valuedomains">Matrix</A> <B>getMatrix</B>()</PRE>
<DL>
<DD>Gets the matrix value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>matrix Scoring (comparison) matrix to use when searching the database.
 Supported matrices inludes: BLOSUM45, BLOSUM50, BLOSUM62, BLOSUM80,
 BLOSUM90, PAM30, PAM70, PAM250.</DL>
</DD>
</DL>
<HR>

<A NAME="setMatrix(uk.org.mygrid.cagrid.valuedomains.Matrix)"><!-- --></A><H3>
setMatrix</H3>
<PRE>
public void <B>setMatrix</B>(<A HREF="../../../../../../uk/org/mygrid/cagrid/valuedomains/Matrix.html" title="class in uk.org.mygrid.cagrid.valuedomains">Matrix</A>&nbsp;matrix)</PRE>
<DL>
<DD>Sets the matrix value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>matrix</CODE> - Scoring (comparison) matrix to use when searching the database.
 Supported matrices inludes: BLOSUM45, BLOSUM50, BLOSUM62, BLOSUM80,
 BLOSUM90, PAM30, PAM70, PAM250.</DL>
</DD>
</DL>
<HR>

<A NAME="getMaxScores()"><!-- --></A><H3>
getMaxScores</H3>
<PRE>
public java.math.BigInteger <B>getMaxScores</B>()</PRE>
<DL>
<DD>Gets the maxScores value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>maxScores Maximum number of scores to output.</DL>
</DD>
</DL>
<HR>

<A NAME="setMaxScores(java.math.BigInteger)"><!-- --></A><H3>
setMaxScores</H3>
<PRE>
public void <B>setMaxScores</B>(java.math.BigInteger&nbsp;maxScores)</PRE>
<DL>
<DD>Sets the maxScores value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>maxScores</CODE> - Maximum number of scores to output.</DL>
</DD>
</DL>
<HR>

<A NAME="getMismatch()"><!-- --></A><H3>
getMismatch</H3>
<PRE>
public java.math.BigInteger <B>getMismatch</B>()</PRE>
<DL>
<DD>Gets the mismatch value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>mismatch Penalty for a nucleotide mismatch. Only when program is set to
 BLASTN.</DL>
</DD>
</DL>
<HR>

<A NAME="setMismatch(java.math.BigInteger)"><!-- --></A><H3>
setMismatch</H3>
<PRE>
public void <B>setMismatch</B>(java.math.BigInteger&nbsp;mismatch)</PRE>
<DL>
<DD>Sets the mismatch value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>mismatch</CODE> - Penalty for a nucleotide mismatch. Only when program is set to
 BLASTN.</DL>
</DD>
</DL>
<HR>

<A NAME="getOpenGap()"><!-- --></A><H3>
getOpenGap</H3>
<PRE>
public java.math.BigInteger <B>getOpenGap</B>()</PRE>
<DL>
<DD>Gets the openGap value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>openGap Penalty for the first residue in a gap, i.e. score taken away for
 the initiation of the gap in sequence or in structure. To make the
 match more significant one can try to make the gap penalty larger.
 It will decrease the number of gaps and if one has good alignment
 without many gaps, its Z-score will be higher.</DL>
</DD>
</DL>
<HR>

<A NAME="setOpenGap(java.math.BigInteger)"><!-- --></A><H3>
setOpenGap</H3>
<PRE>
public void <B>setOpenGap</B>(java.math.BigInteger&nbsp;openGap)</PRE>
<DL>
<DD>Sets the openGap value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>openGap</CODE> - Penalty for the first residue in a gap, i.e. score taken away for
 the initiation of the gap in sequence or in structure. To make the
 match more significant one can try to make the gap penalty larger.
 It will decrease the number of gaps and if one has good alignment
 without many gaps, its Z-score will be higher.</DL>
</DD>
</DL>
<HR>

<A NAME="getQueryDatabase()"><!-- --></A><H3>
getQueryDatabase</H3>
<PRE>
public <A HREF="../../../../../../uk/org/mygrid/cagrid/domain/common/MolecularSequenceDatabase.html" title="class in uk.org.mygrid.cagrid.domain.common">MolecularSequenceDatabase</A> <B>getQueryDatabase</B>()</PRE>
<DL>
<DD>Gets the queryDatabase value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>queryDatabase</DL>
</DD>
</DL>
<HR>

<A NAME="setQueryDatabase(uk.org.mygrid.cagrid.domain.common.MolecularSequenceDatabase)"><!-- --></A><H3>
setQueryDatabase</H3>
<PRE>
public void <B>setQueryDatabase</B>(<A HREF="../../../../../../uk/org/mygrid/cagrid/domain/common/MolecularSequenceDatabase.html" title="class in uk.org.mygrid.cagrid.domain.common">MolecularSequenceDatabase</A>&nbsp;queryDatabase)</PRE>
<DL>
<DD>Sets the queryDatabase value for this NCBIBlastInputParameters.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>queryDatabase</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="equals(java.lang.Object)"><!-- --></A><H3>
equals</H3>
<PRE>
public boolean <B>equals</B>(java.lang.Object&nbsp;obj)</PRE>
<DL>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>equals</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="hashCode()"><!-- --></A><H3>
hashCode</H3>
<PRE>
public int <B>hashCode</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>hashCode</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getTypeDesc()"><!-- --></A><H3>
getTypeDesc</H3>
<PRE>
public static org.apache.axis.description.TypeDesc <B>getTypeDesc</B>()</PRE>
<DL>
<DD>Return type metadata object
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getSerializer(java.lang.String, java.lang.Class, javax.xml.namespace.QName)"><!-- --></A><H3>
getSerializer</H3>
<PRE>
public static org.apache.axis.encoding.Serializer <B>getSerializer</B>(java.lang.String&nbsp;mechType,
                                                                java.lang.Class&nbsp;_javaType,
                                                                javax.xml.namespace.QName&nbsp;_xmlType)</PRE>
<DL>
<DD>Get Custom Serializer
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getDeserializer(java.lang.String, java.lang.Class, javax.xml.namespace.QName)"><!-- --></A><H3>
getDeserializer</H3>
<PRE>
public static org.apache.axis.encoding.Deserializer <B>getDeserializer</B>(java.lang.String&nbsp;mechType,
                                                                    java.lang.Class&nbsp;_javaType,
                                                                    javax.xml.namespace.QName&nbsp;_xmlType)</PRE>
<DL>
<DD>Get Custom Deserializer
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/NCBIBlastInputParameters.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInput.html" title="class in uk.org.mygrid.cagrid.domain.ncbiblast"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../../../uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastJob.html" title="class in uk.org.mygrid.cagrid.domain.ncbiblast"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../../../index.html?uk/org/mygrid/cagrid/domain/ncbiblast/NCBIBlastInputParameters.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="NCBIBlastInputParameters.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
